Updated ensmallen

LucaCappelletti94 opened the pull request:
Refactored library to support ensmallen version 0.4.0.

pnrobinson reviewed (changes requested):
in ``node2vec.py`` i


1) It would be good to replace
```
 optimizer: Union[str, Optimizer] = "nadam",
```
with  
```
optimizer: Union[str, Optimizer] = tf.keras.optimizers.Nadam,
```
for clarity

2) 
  self._negatives_samples = negatives_samples
should be ``negative_samples`` (spelling error)

3)  window_size: 
-- explanation missingh from parameter doc

4) Documentation
  # If there is more than one value per single sample
        if self._get_true_input_length() > 1:
=> This is for CBOW, right? Should we add a comment about this here?

pnrobinson reviewed (changes requested):
2) abstract_node2vec_sequence
spelling errors
 -- Wthever to shuffle the vectors. => "Whether..."
--   Weight on the probability of returning to node coming from
=>   Weight on the probability of returning to the same node the walk just came from
--  This object can be created using the method available from graph (? do you mean available from ensmallen?)
-- samples_number => should be ``sample_number``

pnrobinson commented on embiggen/sequences/abstract_sequence.py:
```diff
@@ -28,9 +29,12 @@ def __init__(
+        self._seed = seed
 
         super().__init__(
             samples_number=samples_number,
```
samples_number should be spelled ``sample_number``

LucaCappelletti94 replied:
Fixed!

pnrobinson commented on embiggen/sequences/abstract_word2vec_sequence.py:
```diff
@@ -49,7 +49,8 @@ def __init__(
+            seed=seed
         )
 
     def on_epoch_end(self):
```
Wthever to shuffle the vectors
spelling error in doc string, should be Whether

LucaCappelletti94 replied:
Fixed!

pnrobinson commented on embiggen/sequences/link_prediction_sequence.py:
```diff
@@ -79,7 +81,7 @@ def __getitem__(self, idx: int) -> Tuple[np.ndarray, np.ndarray]:
+            self._seed + idx + self.elapsed_epochs,
             batch_size=self.batch_size,
             negative_samples=self._negative_samples,
             graph_to_avoid=self._graph_to_avoid,
```
Spelling errors
-- (first line) Keras Sequence for running Neural Netwok on graph link prediction. => "Network"
--  Index corresponding to batch to be rendered. => "rendered" is not quite right here, "to be returned" is better

LucaCappelletti94 replied:
Fixed both.

pnrobinson commented on embiggen/sequences/node2vec_sequence.py:
```diff
@@ -44,16 +44,15 @@ def __getitem__(self, idx: int) -> Tuple[Tuple[np.ndarray, np.ndarray], None]:
             explore_weight=self._explore_weight,
             change_node_type_weight=self._change_node_type_weight,
             change_edge_type_weight=self._change_edge_type_weight,
-            dense_nodes_mapping=self._dense_nodes_mapping
```
spelling -- should be ``dense_node_mapping``

LucaCappelletti94 replied:
Fixed: since this is also a typo within ensmallen, it will break the current branch. It will be fixed in the next version of ensmallen that we should release this afternoon.

pnrobinson commented on embiggen/visualizations/graph_visualizations.py:
```diff
@@ -442,7 +444,7 @@ def plot_edge_types(
+            egde_labels,
             scatter.legend_elements()[0]
         )
         self._clear_axes(axes, "Edge types")
```
Spelling
-- Return given arrats shuffled synchronously. (arrays)
--"""Reset the axes ticks and set the given title.""" => ``the axis ticks``
-- in _set_legend, "handles" should be "handlers" (param name and documentation), handle has a different meaning from handler
-- Documentation of function plot_node_degrees ("Plot common node types of provided graph.") is probably copy paste error?
-- line 426, egde_labels => edge_labels

LucaCappelletti94 replied:
1. Fixed!
2. Within matplotlib the term `axes` is used to reference this type of object, is it a mistake in this context?
3. The term `handles` is the one used within matplotlib.
4. Fixed typo.
5. Fixed typo.

pnrobinson replied:
axis=singular, axes plural, and so normally in English you would say axis ticks. matplotlib uses 'axes' (correctly) in other contexts, e.g., https://matplotlib.org/api/_as_gen/matplotlib.axes.Axes.set_xticks.html

pnrobinson reviewed (changes requested):
The code looks great, it is really elegant. I have made a bunch of suggestions to correct spelling errors.

LucaCappelletti94 commented:
> in `node2vec.py` i
> 
> 1. It would be good to replace
> 
> ```
>  optimizer: Union[str, Optimizer] = "nadam",
> ```
> 
> with
> 
> ```
> optimizer: Union[str, Optimizer] = tf.keras.optimizers.Nadam,
> ```
> 
> for clarity

That would have a different meaning: `"nadam"` in Keras (passed as a string) is interpreted as 'create a new Nadam object with default parameters' and it's how you usually pass it in Keras models. The other version you propose `tf.keras.optimizers.Nadam` would be a class object. If you were to instead use `tf.keras.optimizers.Nadam()` to get an instance of the object, you would get considerable headaches because default parameters that are editable objects (like dictionaries or lists) behave like static objects that are shared among different instances of the function or method containing them.

> 
> 
> self._negatives_samples = negatives_samples
> should be `negative_samples` (spelling error)

Fixed!

> 
> 1. window_size:
>    -- explanation missingh from parameter doc

Added proper documentation for parameter.

> 2. Documentation
> 
> # If there is more than one value per single sample
> ```
>     if self._get_true_input_length() > 1:
> ```
> 
> => This is for CBOW, right? Should we add a comment about this here?

This is generally to choose to build the model as a CBOW model or a SkipGram model. I have extended the comments to try and make its logic more clear.

LucaCappelletti94 commented:
> 1. abstract_node2vec_sequence
>    spelling errors
>    -- Wthever to shuffle the vectors. => "Whether..."
Addressed typo.
>    --   Weight on the probability of returning to node coming from
>    =>   Weight on the probability of returning to the same node the walk just came from
Replaced.
>    --  This object can be created using the method available from graph (? do you mean available from ensmallen?)
Extended comment.
>    -- samples_number => should be `sample_number`
Fixed typo.

vidarmehr commented:
@LucaCappelletti94 Luca, shall we merge pull request?

LucaCappelletti94 commented:
Yes for me we can absolutely proceed, I was waiting for @pnrobinson seal of approval.

vidarmehr commented:
@LucaCappelletti94 Thanks, Luca. I think you addressed all Peter's comments. Is there anything left that needs Peter's approval?

LucaCappelletti94 commented:
Just the pull requested is still missing his 'check', but if you think that we can proceed with everyone's approval I can merge and publish the updated version.

pnrobinson commented:
I think we can move ahead, molte grazie!
